File: android/app/src/main/java/io/conduit/network/ConduitNode.kt
package io.conduit.network

import io.conduit.logging.ConduitLog
import io.conduit.model.DeviceInfo
import io.conduit.model.DeviceType
import io.conduit.model.PairedDevice
import io.conduit.model.Ports
import io.conduit.protocol.Packet
import io.conduit.protocol.PacketType
import io.conduit.storage.AppStore
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket
import java.util.concurrent.ConcurrentHashMap
import kotlin.random.Random

/**
 * The engine tying discovery, connections, pairing and the encrypted session together.
 * The service creates one, wires its callbacks, and calls sendTo / broadcast.
 * Mirrors the .NET ConduitNode.
 */
class ConduitNode(private val store: AppStore) {
    private val log = ConduitLog.tag("Node")
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    private val crypto: CryptoService = CryptoService.loadOrCreate(store.privateKey, store.publicKey).also {
        store.privateKey = it.privateKeyBase64
        store.publicKey = it.publicKeyBase64
    }

    val self = DeviceInfo(
        deviceId = store.deviceId,
        name = store.deviceName,
        type = DeviceType.ANDROID,
        tcpPort = Ports.TCP,
    )

    private val peers = ConcurrentHashMap<String, PeerConnection>()
    private val known = ConcurrentHashMap<String, DeviceInfo>()

    // Devices the user manually disconnected: don't auto-reconnect until they reconnect.
    private val suppressReconnect = ConcurrentHashMap.newKeySet<String>()

    // Devices with an outbound dial already in flight — stops a second beacon from
    // opening a duplicate connection before the first has finished its handshake.
    private val connecting = ConcurrentHashMap.newKeySet<String>()
    private val discovery = DeviceDiscovery(self, ::onBeacon)
    private var serverSocket: ServerSocket? = null

    // Callbacks the UI/service subscribe to.
    var onDevicesChanged: (() -> Unit)? = null
    var onPeerConnected: ((DeviceInfo) -> Unit)? = null
    var onPeerDisconnected: ((DeviceInfo) -> Unit)? = null
    var onPacket: ((DeviceInfo, Packet) -> Unit)? = null
    /** Return true to accept a pairing request. Defaults to accepting (UI shows the code). */
    // The UI shows a confirm/reject dialog, then calls the provided respond() with the decision.
    var onPairingRequest: ((peer: DeviceInfo, code: String, respond: (Boolean) -> Unit) -> Unit)? = null

    val knownDevices: List<DeviceInfo> get() = known.values.toList()

    fun start() {
        // Surface remembered peers up front (as offline) so the user can see and manage them
        // before any beacon arrives — including stale ones left by a peer reinstall.
        for (p in store.pairedDevices()) {
            known.putIfAbsent(
                p.deviceId,
                DeviceInfo(deviceId = p.deviceId, name = p.name, type = p.type, isPaired = true),
            )
        }

        val server = ServerSocket().apply {
            reuseAddress = true
            bind(InetSocketAddress(Ports.TCP))
        }
        serverSocket = server
        scope.launch { acceptLoop(server) }
        discovery.start()
        scope.launch { heartbeatLoop() }
        log.i("Conduit node '${self.name}' started on TCP ${Ports.TCP}")
    }

    // ---- Incoming connections -------------------------------------------------

    private fun acceptLoop(server: ServerSocket) {
        while (scope.isActive) {
            try {
                val client = server.accept()
                log.d("Incoming TCP from ${client.inetAddress?.hostAddress}")
                handleConnection(client)
            } catch (e: Exception) {
                if (scope.isActive) log.w(e, "Accept loop error")
            }
        }
    }

    private fun handleConnection(socket: Socket) {
        val conn = PeerConnection(socket, crypto, self)
        wire(conn)
        conn.start()
    }

    // ---- Outgoing connections -------------------------------------------------

    private fun onBeacon(device: DeviceInfo) {
        val isNew = !known.containsKey(device.deviceId)
        known.compute(device.deviceId) { _, existing ->
            (existing ?: device).apply {
                name = device.name
                ipAddress = device.ipAddress
                tcpPort = device.tcpPort
                lastSeen = System.currentTimeMillis()
                isPaired = store.isPaired(device.deviceId)
            }
        }
        if (isNew) {
            log.i("Discovered ${device.name}")
            onDevicesChanged?.invoke()
        }
        // Only one side dials (the smaller deviceId); the other only accepts. Both sides
        // auto-dialing races to open two sockets at once and can "glare" — each keeps a
        // different one and closes the other — which churns the session and breaks
        // transfers. Explicit/pairing connects bypass this.
        if (store.isPaired(device.deviceId) && !peers.containsKey(device.deviceId) && device.ipAddress != null &&
            !suppressReconnect.contains(device.deviceId) &&
            self.deviceId < device.deviceId
        ) {
            connect(device)
        }
    }

    fun connect(device: DeviceInfo) {
        // An explicit connect clears any manual-disconnect suppression.
        suppressReconnect.remove(device.deviceId)
        if (peers.containsKey(device.deviceId) || device.ipAddress == null) return
        // Coalesce concurrent dials: if one is already in flight for this device, skip.
        if (!connecting.add(device.deviceId)) return
        scope.launch {
            try {
                log.i("Connecting to ${device.name} @ ${device.ipAddress}:${device.tcpPort}")
                val socket = Socket()
                socket.connect(InetSocketAddress(device.ipAddress, device.tcpPort), 5000)
                val conn = PeerConnection(socket, crypto, self)
                wire(conn)
                conn.start()
            } catch (e: Exception) {
                log.w(e, "Failed to connect to ${device.name}")
            } finally {
                connecting.remove(device.deviceId)
            }
        }
    }

    private fun wire(conn: PeerConnection) {
        conn.onHandshaked = { peer ->
            // A completed session means someone reconnected on purpose — allow auto-reconnect again.
            suppressReconnect.remove(peer.deviceId)
            peer.isPaired = store.isPaired(peer.deviceId)

            // Keep exactly one live session per peer. A burst of duplicate connections
            // (both sides dialing, or several beacons before the first handshake lands)
            // would otherwise let a single file's packets race across sockets and truncate
            // session-based transfers. Adopt the first connection to arrive and drop any
            // later duplicate — never the established one, which may be mid-transfer.
            val existing = peers.putIfAbsent(peer.deviceId, conn)
            if (existing != null && existing !== conn) {
                log.i("Duplicate session with ${peer.name}; dropping the redundant connection")
                conn.close()
            } else {
                known[peer.deviceId] = peer
                log.i("Peer connected: ${peer.name} (paired=${peer.isPaired})")
                onPeerConnected?.invoke(peer)
                onDevicesChanged?.invoke()
            }
        }
        conn.onPacket = { packet -> conn.peer?.let { onIncoming(conn, it, packet) } }
        conn.onDisconnected = {
            conn.peer?.let { p ->
                // Only the currently-registered connection owns the peer entry. A superseded
                // duplicate closing must not evict the live session or fire "disconnected".
                if (peers.remove(p.deviceId, conn)) {
                    onPeerDisconnected?.invoke(p)
                    onDevicesChanged?.invoke()
                }
            }
        }
    }

    private fun onIncoming(conn: PeerConnection, peer: DeviceInfo, packet: Packet) {
        when (packet.type) {
            PacketType.PAIR_REQUEST -> handlePairRequest(conn, peer, packet)
            PacketType.PAIR_RESPONSE -> handlePairResponse(peer, packet)
            PacketType.DISCONNECT -> {
                log.i("${peer.name} disconnected")
                suppressReconnect.add(peer.deviceId)
                peers.remove(peer.deviceId)
                onPeerDisconnected?.invoke(peer)
                onDevicesChanged?.invoke()
                conn.close()
            }
            else -> {
                // Security gate: only paired peers may use features. An unpaired peer can still
                // complete the handshake and exchange pair-request/response (handled above), but
                // every feature packet is dropped until it's actually paired.
                if (!store.isPaired(peer.deviceId)) {
                    log.w("Dropping ${packet.type} from unpaired peer ${peer.name}")
                    return
                }
                onPacket?.invoke(peer, packet)
            }
        }
    }

    // ---- Pairing --------------------------------------------------------------

    private fun handlePairRequest(conn: PeerConnection, peer: DeviceInfo, packet: Packet) {
        val code = packet.getString("code") ?: "------"
        val publicKey = packet.getString("publicKey") ?: ""
        log.i("Pair request from ${peer.name}, code $code")

        // Ask the user (via the UI) to confirm the code before accepting. respond() runs when
        // they answer; only then do we trust the peer and reply. No UI wired → reject (the code
        // must be confirmed by a human — never silently auto-accept).
        val respond: (Boolean) -> Unit = { accepted ->
            scope.launch {
                if (accepted && publicKey.isNotEmpty()) {
                    store.addPaired(PairedDevice(peer.deviceId, peer.name, peer.type, publicKey))
                    peer.isPaired = true
                }
                val response = Packet.create(PacketType.PAIR_RESPONSE) {
                    put("accepted", accepted)
                    put("publicKey", crypto.publicKeyBase64)
                }
                if (accepted) {
                    conn.send(response)
                    onDevicesChanged?.invoke()
                } else {
                    // Reject: send the refusal, then drop the session so it doesn't linger as "connected".
                    conn.closeWith(response)
                }
                log.i("Pair request from ${peer.name} ${if (accepted) "accepted" else "rejected"}")
            }
        }
        val handler = onPairingRequest
        if (handler != null) handler(peer, code, respond) else respond(false)
    }

    private fun handlePairResponse(peer: DeviceInfo, packet: Packet) {
        val accepted = packet.getBool("accepted")
        val publicKey = packet.getString("publicKey") ?: ""
        log.i("Pair response from ${peer.name}: accepted=$accepted")
        if (accepted && publicKey.isNotEmpty()) {
            store.addPaired(PairedDevice(peer.deviceId, peer.name, peer.type, publicKey))
            peer.isPaired = true
            onDevicesChanged?.invoke()
        }
    }

    /** Begin pairing with a discovered device; returns the 6-digit code to show the user. */
    suspend fun startPairing(device: DeviceInfo): String {
        if (!peers.containsKey(device.deviceId)) {
            connect(device)
            delay(600)
        }
        val conn = peers[device.deviceId] ?: throw IllegalStateException("Not connected to device")
        val code = Random.nextInt(0, 1_000_000).toString().padStart(6, '0')
        conn.send(Packet.create(PacketType.PAIR_REQUEST) {
            put("publicKey", crypto.publicKeyBase64)
            put("code", code)
        })
        log.i("Started pairing with ${device.name}, code $code")
        return code
    }

    // ---- Sending --------------------------------------------------------------

    fun isConnected(deviceId: String) = peers.containsKey(deviceId)

    /** The last-known IP of a device, for opening a side channel (e.g. the file stream). */
    fun ipFor(deviceId: String): String? = known[deviceId]?.ipAddress

    /**
     * The AES-256 session key shared with a paired peer, derived from its stored public key.
     * Deterministic (ECDH), so the file-stream side channel encrypts with the same key the main
     * session uses, without needing the live connection.
     */
    fun sessionKeyFor(deviceId: String): ByteArray? {
        val pub = store.getPaired(deviceId)?.publicKey ?: return null
        return try { crypto.deriveSessionKey(pub) } catch (e: Exception) { null }
    }

    /**
     * Drop the live session with a device and stop auto-reconnecting to it until the user
     * explicitly connects again (or the app restarts).
     */
    fun disconnect(deviceId: String) {
        suppressReconnect.add(deviceId)
        // Drop the peer from the connected set immediately so the UI flips to "offline"
        // right away, then send the farewell + close the socket in the background.
        val conn = peers.remove(deviceId)
        if (conn != null) {
            log.i("Disconnecting from $deviceId (manual)")
            known[deviceId]?.let { onPeerDisconnected?.invoke(it) }
            onDevicesChanged?.invoke()
            // Tell the peer so it also stops auto-reconnecting, then close.
            conn.closeWith(Packet.create(PacketType.DISCONNECT))
        }
    }

    /**
     * Forget a device completely: drop any live session, delete the stored pairing, and remove
     * it from the device list. Clears out stale entries (e.g. a peer reinstalled under a new
     * identity, leaving the old one orphaned).
     */
    fun forget(deviceId: String) {
        disconnect(deviceId)
        store.removePaired(deviceId)
        known.remove(deviceId)
        suppressReconnect.remove(deviceId)
        log.i("Forgot device $deviceId")
        onDevicesChanged?.invoke()
    }

    fun sendTo(deviceId: String, packet: Packet): Boolean {
        val conn = peers[deviceId]
        if (conn != null) {
            conn.send(packet)
            return true
        }
        log.d("No active connection to $deviceId; dropping ${packet.type}")
        return false
    }

    fun broadcast(packet: Packet) = peers.keys.forEach { sendTo(it, packet) }

    private suspend fun heartbeatLoop() {
        while (scope.isActive) {
            delay(15_000)
            peers.keys.forEach { sendTo(it, Packet.create(PacketType.PING)) }
        }
    }

    fun stop() {
        scope.cancel()
        peers.values.forEach { it.close() }
        discovery.stop()
        runCatching { serverSocket?.close() }
        log.i("Conduit node stopped")
    }
}

File: windows/src/Conduit.Core/Networking/ConduitNode.cs
using System.Collections.Concurrent;
using System.Net;
using System.Net.Sockets;
using Conduit.Core.Logging;
using Conduit.Core.Models;
using Conduit.Core.Protocol;
using Conduit.Core.Security;
using Conduit.Core.Storage;
using Serilog;

namespace Conduit.Core.Networking;

public sealed class PacketEventArgs(DeviceInfo peer, Packet packet) : EventArgs
{
    public DeviceInfo Peer { get; } = peer;
    public Packet Packet { get; } = packet;
}

public sealed class PairingRequestEventArgs(DeviceInfo peer, string code) : EventArgs
{
    public DeviceInfo Peer { get; } = peer;
    public string Code { get; } = code;
    /// <summary>Set by the UI to accept/reject the pairing.</summary>
    public bool Accepted { get; set; }
}

/// <summary>
/// The engine that ties discovery, connections, pairing and the encrypted session together.
/// The UI layer creates one of these, subscribes to its events, and calls SendToAsync.
/// </summary>
public sealed class ConduitNode : IAsyncDisposable
{
    private readonly ILogger _log = ConduitLog.For("Node");
    private readonly AppStore _store;
    private readonly CryptoService _crypto;
    private readonly DeviceInfo _self;
    private readonly DeviceDiscovery _discovery;
    private readonly ConcurrentDictionary<string, PeerConnection> _peers = new();
    private readonly ConcurrentDictionary<string, DeviceInfo> _known = new();

    // Devices the user manually disconnected: don't auto-reconnect until they reconnect.
    private readonly ConcurrentDictionary<string, byte> _suppressReconnect = new();

    private TcpListener? _listener;
    private CancellationTokenSource? _cts;
    private Task? _acceptTask;
    private Task? _heartbeatTask;

    public IReadOnlyCollection<DeviceInfo> KnownDevices => _known.Values.ToList();
    public DeviceInfo Self => _self;

    public event EventHandler? DevicesChanged;
    public event EventHandler<DeviceInfo>? PeerConnected;
    public event EventHandler<DeviceInfo>? PeerDisconnected;
    public event EventHandler<PacketEventArgs>? PacketReceived;
    public event EventHandler<PairingRequestEventArgs>? PairingRequested;

    public ConduitNode(AppStore store)
    {
        _store = store;
        _crypto = CryptoService.LoadOrCreate(store.Config.PrivateKey, out var privateKey);
        if (store.Config.PrivateKey != privateKey)
        {
            store.Config.PrivateKey = privateKey;
            store.Save();
        }

        _self = new DeviceInfo
        {
            DeviceId = store.Config.DeviceId,
            Name = store.Config.DeviceName,
            Type = DeviceType.Windows,
            TcpPort = ConduitPorts.Tcp
        };
        _discovery = new DeviceDiscovery(_self);
        _discovery.BeaconReceived += OnBeaconReceived;
    }

    public async Task StartAsync()
    {
        _cts = new CancellationTokenSource();

        _listener = new TcpListener(IPAddress.Any, ConduitPorts.Tcp);
        _listener.Start();
        _acceptTask = Task.Run(() => AcceptLoopAsync(_cts.Token));

        _discovery.Start();
        await _discovery.AnnounceAsync().ConfigureAwait(false);

        _heartbeatTask = Task.Run(() => HeartbeatLoopAsync(_cts.Token));
        _log.Information("Conduit node '{Name}' started on TCP {Port}", _self.Name, ConduitPorts.Tcp);
    }

    // ---- Incoming connections -------------------------------------------------

    private async Task AcceptLoopAsync(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            try
            {
                var client = await _listener!.AcceptTcpClientAsync(ct).ConfigureAwait(false);
                _log.Debug("Incoming TCP from {Remote}", client.Client.RemoteEndPoint);
                _ = HandleConnectionAsync(client, ct);
            }
            catch (OperationCanceledException) { break; }
            catch (Exception ex)
            {
                _log.Warning(ex, "Accept loop error");
            }
        }
    }

    private async Task HandleConnectionAsync(TcpClient client, CancellationToken ct)
    {
        var conn = new PeerConnection(client, _crypto, _self);
        WireConnection(conn);
        await conn.RunAsync(ct).ConfigureAwait(false);
    }

    // ---- Outgoing connections (triggered by beacons) --------------------------

    private void OnBeaconReceived(object? sender, BeaconEventArgs e)
    {
        var device = e.Device;
        bool isNew = !_known.ContainsKey(device.DeviceId);
        _known.AddOrUpdate(device.DeviceId, device, (_, existing) =>
        {
            existing.Name = device.Name;
            existing.IpAddress = device.IpAddress;
            existing.TcpPort = device.TcpPort;
            existing.LastSeen = DateTimeOffset.UtcNow;
            return existing;
        });
        device.IsPaired = _store.IsPaired(device.DeviceId);

        if (isNew)
        {
            _log.Information("Discovered {Device}", device);
            DevicesChanged?.Invoke(this, EventArgs.Empty);
        }

        // Auto-connect to paired peers that aren't connected yet — unless the user
        // manually disconnected this one. Only one side dials (the smaller DeviceId);
        // the other only accepts. Both sides auto-dialing races to open two sockets at
        // once and can "glare" — each keeps a different one and closes the other — which
        // churns the session and breaks transfers. Explicit/pairing connects bypass this.
        if (device.IsPaired && !_peers.ContainsKey(device.DeviceId) && device.IpAddress is not null
            && !_suppressReconnect.ContainsKey(device.DeviceId)
            && string.CompareOrdinal(_self.DeviceId, device.DeviceId) < 0)
            _ = ConnectAsync(device);
    }

    public async Task ConnectAsync(DeviceInfo device)
    {
        // An explicit connect clears any manual-disconnect suppression.
        _suppressReconnect.TryRemove(device.DeviceId, out _);
        if (_peers.ContainsKey(device.DeviceId) || device.IpAddress is null) return;
        try
        {
            _log.Information("Connecting to {Device} @ {Ip}:{Port}", device, device.IpAddress, device.TcpPort);
            var client = new TcpClient();
            await client.ConnectAsync(device.IpAddress, device.TcpPort).ConfigureAwait(false);
            var conn = new PeerConnection(client, _crypto, _self);
            WireConnection(conn);
            _ = conn.RunAsync(_cts!.Token);
        }
        catch (Exception ex)
        {
            _log.Warning(ex, "Failed to connect to {Device}", device);
        }
    }

    private void WireConnection(PeerConnection conn)
    {
        conn.Handshaked += (_, peer) =>
        {
            // A completed session means someone reconnected on purpose — allow auto-reconnect again.
            _suppressReconnect.TryRemove(peer.DeviceId, out byte _);
            peer.IsPaired = _store.IsPaired(peer.DeviceId);

            // Keep exactly one live session per peer. A burst of duplicate connections
            // (both sides dialing, or several beacons before the first handshake lands)
            // would otherwise let a single file's packets race across sockets and
            // truncate session-based transfers. Adopt the first connection to arrive and
            // drop any later duplicate — never the established one, which may be mid-transfer.
            if (!_peers.TryAdd(peer.DeviceId, conn))
            {
                _log.Information("Duplicate session with {Peer}; dropping the redundant connection", peer);
                _ = conn.DisposeAsync();
                return;
            }

            _known[peer.DeviceId] = peer;
            _log.Information("Peer connected: {Peer} (paired={Paired})", peer, peer.IsPaired);
            PeerConnected?.Invoke(this, peer);
            DevicesChanged?.Invoke(this, EventArgs.Empty);
        };
        conn.PacketReceived += (_, packet) => OnPacket(conn, packet);
        conn.Disconnected += (_, _) =>
        {
            if (conn.Peer is not { } p) return;

            // Only the currently-registered connection owns the peer entry. A superseded
            // duplicate that closes must not evict the live session or fire "disconnected".
            bool wasCurrent = ((ICollection<KeyValuePair<string, PeerConnection>>)_peers)
                .Remove(new KeyValuePair<string, PeerConnection>(p.DeviceId, conn));
            if (!wasCurrent) return;

            p.State = ConnectionState.Disconnected;
            PeerDisconnected?.Invoke(this, p);
            DevicesChanged?.Invoke(this, EventArgs.Empty);
        };
    }

    private void OnPacket(PeerConnection conn, Packet packet)
    {
        if (conn.Peer is not { } peer) return;

        switch (packet.Type)
        {
            case PacketType.PairRequest:
                HandlePairRequest(conn, peer, packet);
                return;
            case PacketType.PairResponse:
                HandlePairResponse(peer, packet);
                return;
            case PacketType.Disconnect:
                _log.Information("{Peer} disconnected", peer);
                _suppressReconnect[peer.DeviceId] = 1;
                _ = conn.DisposeAsync();
                return;
            default:
                // Security gate: only paired peers may use features. An unpaired peer can still
                // complete the handshake and exchange pair-request/response (handled above), but
                // every feature packet is dropped until it's actually paired.
                if (!_store.IsPaired(peer.DeviceId))
                {
                    _log.Warning("Dropping {Type} from unpaired peer {Peer}", packet.Type, peer);
                    return;
                }
                PacketReceived?.Invoke(this, new PacketEventArgs(peer, packet));
                return;
        }
    }

    // ---- Pairing --------------------------------------------------------------

    private void HandlePairRequest(PeerConnection conn, DeviceInfo peer, Packet packet)
    {
        string code = packet.GetString("code") ?? "------";
        string publicKey = packet.GetString("publicKey") ?? "";
        _log.Information("Pair request from {Peer}, code {Code}", peer, code);

        var args = new PairingRequestEventArgs(peer, code);
        PairingRequested?.Invoke(this, args);

        bool accepted = args.Accepted || _store.Config.AutoAcceptFromPaired && _store.IsPaired(peer.DeviceId);
        if (accepted && !string.IsNullOrEmpty(publicKey))
        {
            _store.AddPaired(new PairedDevice
            {
                DeviceId = peer.DeviceId, Name = peer.Name, Type = peer.Type, PublicKey = publicKey
            });
            peer.IsPaired = true;
        }

        var response = Packet.Create(PacketType.PairResponse, b =>
        {
            b["accepted"] = accepted;
            b["publicKey"] = _crypto.PublicKeyBase64;
        });
        if (accepted)
        {
            _ = conn.SendAsync(response);
            DevicesChanged?.Invoke(this, EventArgs.Empty);
        }
        else
        {
            // Reject: send the refusal, then drop the session so it doesn't linger as "connected".
            _ = Task.Run(async () =>
            {
                try { await conn.SendAsync(response); } catch { /* best effort */ }
                await conn.DisposeAsync();
            });
        }
    }

    private void HandlePairResponse(DeviceInfo peer, Packet packet)
    {
        bool accepted = packet.GetBool("accepted");
        string publicKey = packet.GetString("publicKey") ?? "";
        _log.Information("Pair response from {Peer}: accepted={Accepted}", peer, accepted);
        if (accepted && !string.IsNullOrEmpty(publicKey))
        {
            _store.AddPaired(new PairedDevice
            {
                DeviceId = peer.DeviceId, Name = peer.Name, Type = peer.Type, PublicKey = publicKey
            });
            peer.IsPaired = true;
            DevicesChanged?.Invoke(this, EventArgs.Empty);
        }
    }

    /// <summary>Begin pairing with a discovered device. Returns the 6-digit code to show the user.</summary>
    public async Task<string> StartPairingAsync(DeviceInfo device)
    {
        if (!_peers.TryGetValue(device.DeviceId, out var conn))
        {
            await ConnectAsync(device).ConfigureAwait(false);
            await Task.Delay(500).ConfigureAwait(false);
            _peers.TryGetValue(device.DeviceId, out conn);
        }
        if (conn is null) throw new InvalidOperationException("Not connected to device");

        string code = Random.Shared.Next(0, 1_000_000).ToString("D6");
        await conn.SendAsync(Packet.Create(PacketType.PairRequest, b =>
        {
            b["publicKey"] = _crypto.PublicKeyBase64;
            b["code"] = code;
        })).ConfigureAwait(false);
        _log.Information("Started pairing with {Device}, code {Code}", device, code);
        return code;
    }

    // ---- Sending --------------------------------------------------------------

    public bool IsConnected(string deviceId) => _peers.ContainsKey(deviceId);

    /// <summary>The last-known IP of a device, for opening a side channel (e.g. the file stream).</summary>
    public string? IpFor(string deviceId) =>
        _known.TryGetValue(deviceId, out var d) ? d.IpAddress : null;

    /// <summary>
    /// The AES-256 session key shared with a paired peer, derived from its stored public key.
    /// It's deterministic (ECDH), so the file-stream side channel can encrypt with the same key
    /// the main session uses, without holding a reference to the live connection.
    /// </summary>
    public byte[]? SessionKeyFor(string deviceId)
    {
        var pub = _store.GetPaired(deviceId)?.PublicKey;
        if (string.IsNullOrEmpty(pub)) return null;
        try { return _crypto.DeriveSessionKey(pub); } catch { return null; }
    }

    /// <summary>
    /// Drop the live session with a device and stop auto-reconnecting to it until the user
    /// explicitly connects again (or the app restarts).
    /// </summary>
    public async Task DisconnectAsync(string deviceId)
    {
        _suppressReconnect[deviceId] = 1;
        if (_peers.TryGetValue(deviceId, out var conn))
        {
            _log.Information("Disconnecting from {Id} (manual)", deviceId);
            // Tell the peer so it also stops auto-reconnecting, then close.
            try { await conn.SendAsync(Packet.Create(PacketType.Disconnect)).ConfigureAwait(false); }
            catch (Exception ex) { _log.Debug(ex, "Could not send disconnect notice to {Id}", deviceId); }
            await conn.DisposeAsync().ConfigureAwait(false);
        }
    }

    public async Task<bool> SendToAsync(string deviceId, Packet packet)
    {
        if (_peers.TryGetValue(deviceId, out var conn))
        {
            try
            {
                await conn.SendAsync(packet).ConfigureAwait(false);
                return true;
            }
            catch (Exception ex)
            {
                _log.Warning(ex, "Send to {Id} failed", deviceId);
            }
        }
        else
        {
            _log.Debug("No active connection to {Id}; dropping {Type}", deviceId, packet.Type);
        }
        return false;
    }

    public async Task BroadcastAsync(Packet packet)
    {
        foreach (var id in _peers.Keys)
            await SendToAsync(id, packet).ConfigureAwait(false);
    }

    // ---- Heartbeat ------------------------------------------------------------

    private async Task HeartbeatLoopAsync(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            try { await Task.Delay(TimeSpan.FromSeconds(15), ct).ConfigureAwait(false); }
            catch (OperationCanceledException) { break; }

            foreach (var id in _peers.Keys)
                await SendToAsync(id, Packet.Create(PacketType.Ping)).ConfigureAwait(false);
        }
    }

    public async ValueTask DisposeAsync()
    {
        _cts?.Cancel();
        foreach (var conn in _peers.Values)
            await conn.DisposeAsync().ConfigureAwait(false);
        _discovery.Dispose();
        _listener?.Stop();
        _cts?.Dispose();
        _log.Information("Conduit node stopped");
    }
}
